File: src/MGRAST/bin/test_cass.py
#!/usr/bin/env python

import sys
import time
import random
import psycopg2
import requests
from optparse import OptionParser
from cassandra.cluster import Cluster
from cassandra.policies import RetryPolicy

MAX_INT = 24468843

def random_array(size):
    array = []
    for i in range(size):
        array.append(random.randint(1, MAX_INT))
    return array

def main(args):
    parser = OptionParser()
    parser.add_option("", "--batch", dest="batch", type="int", default=100, help="query batch size, default: 100")
    parser.add_option("", "--count", dest="count", type="int", default=10, help="number of query iterations, default: 10")
    parser.add_option("", "--phost", dest="phost", default=None, help="Postgres host")
    parser.add_option("", "--pname", dest="pname", default=None, help="Postgres name")
    parser.add_option("", "--puser", dest="puser", default=None, help="Postgres user")
    parser.add_option("", "--chost", dest="chost", default=None, help="Cassandra host")
    parser.add_option("", "--cname", dest="cname", default=None, help="Cassandra name")
    parser.add_option("", "--shost", dest="shost", default=None, help="SOLR host")
    parser.add_option("", "--sname", dest="sname", default=None, help="SOLR collection")
    (opts, args) = parser.parse_args()
    
    handle = None
    mode = ""
    
    # test PSQL
    if opts.phost and opts.pname and opts.puser:
        mode = "sql"
        handle = psycopg2.connect(
            host=opts.phost,
            database=opts.pname,
            user=opts.puser
        )
    # test CASS
    elif opts.chost and opts.cname:
        mode = "cql"
        handle = Cluster(
            contact_points=[opts.chost],
            default_retry_policy = RetryPolicy()
        )
    # test solr
    elif opts.shost and opts.sname:
        mode = "solr"
        surl = "http://"+opts.shost+"/solr/"+opts.sname+"/select"
        shead = {'Content-Type': 'application/x-www-form-urlencoded'}
        sfields = "%2C".join(['md5_id', 'source', 'md5', 'accession', 'function', 'organism'])
    else:
        parser.error("Invalid usage")
    
    start = time.time()
    found = set()
    for i in range(opts.count):
        ints = random_array(opts.batch)
        query = "SELECT * FROM md5_id_annotation WHERE id IN ("+",".join(map(str, ints))+");"
        if mode == "sql":
            cursor = handle.cursor()
            cursor.execute(query)
            data = cursor.fetchone()
            while (data):
                found.add(data[2])
                data = cursor.fetchone()
        elif mode == "cql":
            session = handle.connect(opts.cname)
            rows = session.execute(query)
            for r in rows:
                found.add(r.md5)
        elif mode == "solr":
            query = "md5_id:("+" OR ".join(map(str, ints))+")"
            sdata = "q=*%3A*&fq="+query+"&start=0&rows=1000000000&wt=json&fl="+sfields
            req = requests.post(surl, headers=shead, data=sdata, allow_redirects=True)
            rj = req.json()
            for d in rj['response']['docs']:
                found.add(d['md5'])
    
    end = time.time()
    print "%d loops of size %d ran in %d seconds"%(opts.count, opts.batch, int(round(end-start)))
    print "%d ids requested, %d md5s found"%((opts.count * opts.batch), len(found))
    

if __name__ == "__main__":
    sys.exit( main(sys.argv) )

File: src/MGRAST/bin/test_cass.pl
#!/usr/bin/env perl

use strict;
use warnings;

use JSON;
use Getopt::Long;
use Data::Dumper;
use Devel::Size qw(size total_size);
use Inline::Python qw(py_eval);

my $batch = 1000;
my $host = "";
my $name = "";
my $usage = qq($0
  --batch  query batch size, default: 1000
  --host   Cassandra host
  --name   Cassandra name
);

if ( (@ARGV > 0) && ($ARGV[0] =~ /-h/) ) { print STDERR $usage; exit 1; }
if ( ! GetOptions(
    'batch:i' => \$batch,
	'host:s'  => \$host,
	'name:s'  => \$name
   ) ) {
  print STDERR $usage; exit 1;
}

unless ($host && $name) {
    print STDERR $usage; exit 1;
}

my $python = q(
import random
from cassandra.cluster import Cluster
from cassandra.policies import RetryPolicy
from cassandra.query import dict_factory

class TestCass(object):
    def __init__(self, host, name):
        self.timeout = 300
        self.max_int = 24468843
        self.name = name
        self.hosts = host.split(",")
        self.handle = Cluster(
            contact_points = self.hosts,
            default_retry_policy = RetryPolicy()
        )
        self.session = self.handle.connect(self.name)
        self.session.default_timeout = self.timeout
        self.session.row_factory = dict_factory
        #self.prep = self.session.prepare("SELECT * FROM id_annotation WHERE id IN ? AND source=?")
    def random_array(self, size):
        array = []
        for i in range(size):
            array.append(random.randint(1, self.max_int))
        return array
    def get_records(self, ids, source='RefSeq'):
        found = []
        query = "SELECT * FROM id_annotation WHERE id IN (%s) AND source='%s'"%(",".join(map(str, ids)), source)
        rows = self.session.execute(query)
        for r in rows:
            found.append(r)
        return found
    def get_md5s(self, ids, source='RefSeq'):
        found = []
        query = "SELECT * FROM id_annotation WHERE id IN (%s) AND source='%s'"%(",".join(map(str, ids)), source)
        rows = self.session.execute(query)
        for r in rows:
            found.append(r["md5"])
        return found
    def get_iter(self, ids, source='RefSeq'):
        query = "SELECT * FROM id_annotation WHERE id IN (%s) AND source='%s'"%(",".join(map(str, ids)), source)
        return self.session.execute(query)
);

py_eval($python);

my $start = time;
my $json = JSON->new;
$json = $json->utf8();
$json->max_size(0);
$json->allow_nonref;

my $tester = Inline::Python::Object->new('__main__', 'TestCass', $host, $name);
my $fields = join('%2C', ('md5_id', 'source', 'md5', 'accession', 'function', 'organism'));

my $random_ids = $tester->random_array($batch);
my $recs = $tester->get_records($random_ids);
my $md5s = $tester->get_md5s($random_ids);
my $iter = $tester->get_iter($random_ids);

print STDERR "\n";
my $end = time;

print "time: ".($end - $start)." seconds\n";
print "recs size: ".total_size($recs)." bytes\n";
print "md5s size: ".total_size($md5s)." bytes\n";
print "iter size: ".total_size($iter)." bytes\n";
print "recs found: ".scalar(@$recs)."\n";
print "md5s found: ".scalar(@$md5s)."\n";
print "first recs: ".Dumper($recs->[0])."\n";
print "first md5s: ".Dumper($md5s->[0])."\n";
print "iter: ".Dumper($iter)."\n";

